### PR TITLE
Fix fix develop

### PR DESCRIPTION
**Pull request by sharabash**
@mmahalwy please review and merge

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,44 +1,42 @@
...
-[![Stories In Progress](https://badge.waffle.io/quran/quran.com-frontend.svg?label=in%20progress&title=In%20Progress)](http://waffle.io/quran/quran.com-frontend)
-[![Stories In Review](https://badge.waffle.io/quran/quran.com-frontend.svg?label=in%20review&title=In%20Review)](http://waffle.io/quran/quran.com-frontend)
-
-
-## Quran.com
-This is the project soon to be the Quran.com facing site. This is built in
-Reactjs + Redux + Expressjs + Webpack. It is isomorphic (javascript shared
-between both the server and the client) for SEO reasons.
+# quran.com
+This is the project soon to be the quran.com facing site. This is built in
```
Comment (mmahalwy) [inline README.md]: ✂️

File: README.md [hunk 2]
```diff
@@ -1,44 +1,42 @@
...
+Reactjs + Redux + Expressjs + Webpack. It is isomorphic which is fancy speak
```
Comment (mmahalwy) [inline README.md]: I think the way it was was fine

File: README.md [hunk 3]
```diff
@@ -1,44 +1,42 @@
...
 
-#### Getting started
-Simply clone this repo, then run `npm install` to install all the required node_modules.
-From there, you are ready to go!
 
-#### Developing
-To start the app, run `npm run dev` which will
-run both the server and the client (webpack) to compile upon edits. Go to localhost:8001 in your browser, not 8000 (that is just the express server).
+### Getting started
+If you hate reading then skip to the 'Lazy synopsis' sub-section. If you've got 5 minutes to spare, read this entire section.
```
Comment (mmahalwy) [inline README.md]: Why not just call it `TL;DR` and have it here and let the reader continue if they need to?

File: README.md [hunk 4]
```diff
@@ -1,44 +1,42 @@
...
+#### Getting started -- Installation
+Simply clone this repo, then run `npm install` to install all the required node_modules. You should be running a relatively recent version
+of Node.js, around ~5.0. If you have an archaic version of Node.js installed and are dreading the prospect of upgrading, have no fear, `nvm`
+will save you time and headache -- install it, then use it to manage your version: https://github.com/creationix/nvm
 
-We also have nightwatch function tests. You can install nightwatch globally and can run tests like this:
-```
-nightwatch --test tests/functional/specs/Index_spec.js
-```
+Next, in the local file `./development.env`, an env variable is defined `API_URL`. Make sure this is set to quran.com at port 3000, like this:
```
Comment (mmahalwy) [inline README.md]: `.env` only now

File: README.md [hunk 5]
```diff
@@ -48,4 +46,27 @@ subl bundle-stats.json #so that you can the output
...
+#### How to contribute
+Fork this repo, then create a PR for specific fixes, improvements, etc. Note that wrapping your
+head around things like React and Redux takes time, but all you essentially need to know is javascript
+and have a bit of determination. We trust that you will not plagiarize this project, it is at the end
+of the day for the sake of Allah and we all have good intentions while working with this project,
+but if you're doing something unique, then have at it. We do stress however that "stealing"
+this is unacceptable (e.g. putting up a clone of the website but with ads).
+
+#### Testing - SECTION DEPRECATED
+This section is now deprecated because neither of these commands work. @mmahalwy, please share your knowledge...
```
Comment (mmahalwy) [inline README.md]: `npm test` should work. As for functional I disabled it for now.

File: src/components/LazyLoad/index.js [hunk 6]
```diff
@@ -20,6 +20,19 @@ export default class LazyLoad extends Component {
     offset: 1000
   }
 
+  state = {
+    lazyLoadCalls: 0
+  };
+
+  constructor(){
+    super(...arguments);
+    this.debug = debug('component:LazyLoad');
```
Comment (mmahalwy) [inline src/components/LazyLoad/index.js]: Don't need to set it on `this` thats bad practice. It's a standalone library that should not be saved to a component.

File: src/components/LazyLoad/index.js [hunk 7]
```diff
@@ -20,6 +20,19 @@ export default class LazyLoad extends Component {
     offset: 1000
   }
 
+  state = {
+    lazyLoadCalls: 0
```
Comment (mmahalwy) [inline src/components/LazyLoad/index.js]: why?

File: src/containers/Surah/Title/index.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ const zeroPad = (num, places) => {
 };
 
 const Title = ({ surah }) => {
-  const title = require('../../../../static/images/titles/' + zeroPad(surah.id, 3) + '.svg');
+  const title = '/images/titles/'+ zeroPad(surah.id, 3) +'.svg';
```
Comment (mmahalwy) [inline src/containers/Surah/Title/index.js]: will not work with CDN.

File: src/containers/Surah/Title/index.js [hunk 9]
```diff
@@ -28,7 +28,7 @@ const Title = ({ surah }) => {
       <div className="col-md-6 col-xs-6 surah-title text-center">
         {
           surah &&
-          <img src={title} className="title hidden-xs" alt={`${surah.name.simple} (${surah.name.english}) - سورة ${surah.name.arabic}`} />
+          <a href="/"><img src={title} className="title hidden-xs" alt={`${surah.name.simple} (${surah.name.english}) - سورة ${surah.name.arabic}`} /></a>
```
Comment (mmahalwy) [inline src/containers/Surah/Title/index.js]: nice!

File: src/containers/Surah/index.js [hunk 10]
```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
...
+// then(spinner stop)
+// catch(spinner stop then visual feedback about error)
 @asyncConnect([
   {
-    promise({ store: { getState, dispatch } }) {
-      if (!isAllLoaded(getState())) {
-        return dispatch(loadAll());
-      }
+    promise: function({ store: { getState, dispatch }, params }) { // this is much less confusing then the old line 'promise({ store: { getState, dispatch }, params }) {'-- it looks like a method call even though it's a method definition, misleading trickery
+      this.debug = debug(`container:Surah:asyncConnect ${how_many_times}`);
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: Bad practice, see comment above.

File: src/containers/Surah/index.js [hunk 11]
```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
 let lastScroll = 0;
 const ayahRangeSize = 30;
 
+var how_many_times = 0;
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: no `var`s. and camelCase!

Comment (mmahalwy) [inline src/containers/Surah/index.js]: And this is not descriptive

File: src/containers/Surah/index.js [hunk 12]
```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
...
+// catch(spinner stop then visual feedback about error)
 @asyncConnect([
   {
-    promise({ store: { getState, dispatch } }) {
-      if (!isAllLoaded(getState())) {
-        return dispatch(loadAll());
-      }
+    promise: function({ store: { getState, dispatch }, params }) { // this is much less confusing then the old line 'promise({ store: { getState, dispatch }, params }) {'-- it looks like a method call even though it's a method definition, misleading trickery
+      this.debug = debug(`container:Surah:asyncConnect ${how_many_times}`);
+      how_many_times += 1;
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: Why?

File: src/containers/Surah/index.js [hunk 13]
```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
...
-
-      if (isNaN(surahId)) {
-        // Should have an alert or something to tell user there is an error.
-        return dispatch(push('/'));
-      }
-
-      if (params.surahId !== getState().surahs.current) {
-        dispatch(setCurrentSurah(surahId));
-      }
+      const p = () => {
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: bad variable name

File: src/containers/Surah/index.js [hunk 14]
```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
...
+      };
+
+      this.debug('first promise start');
+      return p()
+        .then((args) => {
+          this.debug(`first promise end ${args}`);
+        })
+        .catch((err) => {
+          this.debug(`first promise err ${err}`);
+        });
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: This whole method is a mess and really difficult to understand what is happening...

File: src/containers/Surah/index.js [hunk 15]
```diff
@@ -133,6 +188,8 @@ const ayahRangeSize = 30;
 export default class Surah extends Component {
   constructor() {
     super(...arguments);
+    this.debug = debug('container:Surah');
+    this.debug('constructor');
```
Comment (mmahalwy) [inline src/containers/Surah/index.js]: bad

File: src/helpers/debug.js [hunk 16]
```diff
@@ -1,9 +1,19 @@
...
+      console.debug ? console.debug(...args) : console.log(...args);
+    };
+    _debug.err = (errs) => {
+      errs = [...(typeof(errs) === 'string' ? [errs] : errs)];
+      if (__CLIENT__) errs.unshift(`quran:${namespace}`);
+      console.error ? console.error(...errs) : console.warn(...errs);
+    };
+    return typeof(args) === 'undefined' ? _debug : _debug(...(typeof(args) === 'string' ? [args] : args));
+  } catch(e) {
+    console.warn('debug failed', namespace, args, e);
```
Comment (mmahalwy) [inline src/helpers/debug.js]: Why change this file?

File: webpack.config.js [hunk 17]
```diff
@@ -10,6 +10,8 @@ var webpackIsomorphicToolsPlugin = new IsomorphicPlugin(require('./webpack-isomo
 var babelrc = fs.readFileSync('./.babelrc');
 var babelrcObject = {};
 
+Object.assign = require('object-assign');
```
Comment (mmahalwy) [inline webpack.config.js]: you dont need this.

File: start.js [hunk 18]
```diff
@@ -8,15 +8,17 @@ global.__CLIENT__ = false;
 global.__SERVER__ = true;
 global.__DEVELOPMENT__ = process.env.NODE_ENV !== 'production';
 
-if (__DEVELOPMENT__) {
-  if (!require('piping')({
-      hook: true,
-      ignore: /(\/\.|~$|\.json|\.scss$)/i
-    })) {
-    return;
+(function() {
```
Comment (mmahalwy) [inline start.js]: why is this in a closure?

**Comment by mmahalwy** [comment]
Also, tests are broken cause you changed the way `<LazyLoad />` works...

**Comment by azizur** [comment]
@sharabash `Fix fix develop` is a really ambiguous title. What are you fixing?

**Comment by mmahalwy** [comment]
@sharabash lets close this for now. feel free to reopen

**Pull request closed without merging**